feat: Editorial review, Relative color syntax (#32004)
mdn/content: files/en-us/web/css/color_value/color/index.md

slug: Web/CSS/color_value/color
page-type: css-function
browser-compat: css.types.color.color
spec-urls:
  - https://drafts.csswg.org/css-color-5/#color-function
  - https://drafts.csswg.org/css-color-5/#relative-color-function
  - https://drafts.csswg.org/css-color/#color-function
---

{{CSSRef}}

> The **`color()`** functional notation allows a color to be specified in a particular, specified {{glossary("color_spaces", "color space")}} rather than the implicit sRGB color space that most of the other color functions operate in.

Reviewer: ```suggestion
The **`color()`** functional notation allows a color to be specified in a particular, specified {{glossary("color space")}} rather than the implicit sRGB color space that most of the other color functions operate in.
```
Works below. Should work here too ;).
Yay. So glad we have that entry. Thanks for the review.

  - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` explicitly specifies no alpha channel. If included, the value is preceded by a slash (`/`).

> **Note:** See [Missing color components](/en-US/docs/Web/CSS/color_value#missing_color_components) for more information on the effect of `none`.

#### Relative value syntax

```text
color(from <color> colorspace c1 c2 c3[ / A])
```

The parameters are as follows:

- `from <color>`
  - : The keyword `from` is always included when defining a relative color, followed by a {{cssxref("&lt;color&gt;")}} value representing the **origin color**. This is the original color that the relative color is based on. Note that the origin color can be _any_ valid {{cssxref("&lt;color&gt;")}} syntax, including another relative color.
- `colorspace`
  - : An {{CSSXref("&lt;ident&gt;")}} denoting the {{glossary("color space")}} of the output color, generally one of the predefined color spaces: `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, `rec2020`, `xyz`, `xyz-d50`, or `xyz-d65`.
- `c1`, `c2`, `c3`

>   - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none` (equivalent to `0` in this case). These values represent the component values for the output color. When using a `<number>` value, generally `0` to `1` represents the bounds of the color space. Values outside of that range are permitted but will be out of {{glossary("Gamut", "gamut")}} for the given color space. Generally, when using a percentage value, `100%` represents `1` and `0%` represents `0`.

Reviewer: ```suggestion
  - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none` (equivalent to `0` in this case). These values represent the component values for the output color. When using a `<number>` value, generally `0` to `1` represents the bounds of the color space. Values outside of that range are permitted but will be out of {{glossary("gamut")}} for the given color space. Generally, when using a percentage value, `100%` represents `1` and `0%` represents `0`.
```
lower case should work.

- `A` {{optional_inline}}

>   - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` explicitly specifies no alpha channel. This represents the alpha channel value of the output color. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).

Reviewer: ```suggestion
  - : An {{CSSXref("&lt;alpha-value&gt;")}}, where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` to explicitly specify no alpha channel. This represents the alpha channel value of the output color. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).
```

Author: I can see why you've suggested this, but it doesn't quite work with the structure and rhythm of the sentence. Which was weird to begin with. To satisfy both our needs on this front, I've modified this on all pages to "...or the keyword `none` can be used to explicitly specify..."

#### Defining relative color output channel components

When using relative color syntax inside a `color()` function, the browser converts the origin color into an equivalent color in the specified color space (if it is not already specified as such). The color is defined as three distinct color channel values plus an alpha channel value (`alpha`). These channel values are made available inside the function to be used when defining the output color channel values:

> - The three color channel values of the origin color are resolved to a `<number>`. For predefined color spaces, depending on which is specified, these values will be one of the following:

Reviewer: ```suggestion
- The three color channel values of the origin color, and the `alpha` value, are resolved to a `<number>`. For predefined color spaces, depending on which is specified, these values will be one of the following:
```
another option for introducing alpha

Author: This line is just about the three color channel values, not the alpha channel, so it makes no sense to mention it here. No change made.

  - `r`, `g`, and `b`: Color channel values for the RGB-based color spaces `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, and `rec2020`.
  - `x`, `y`, and `z`: Color channel values for the CIE XYZ-based color spaces `xyz`, `xyz-d50`, and `xyz-d65`.

> 

Reviewer: ```suggestion
- `alpha`:  In all color spaces, `alpha` is a `<number>` that corresponds to the origin color’s alpha transparency. 

```
We use alpha below, but never defined it. I think this is the right spot for it.

Author: We don't need this line. We define the alpha channel below the two notes. I've updated the style of that line to turn it into more of a definition in the style of the preceding lines:

> - `alpha`: The color's transparency value, resolved to a `<number>` between `0` and `1`, inclusive.

    > **Note:** Each of these values is usually between `0` and `1` but, as explained above, they may be outside these bounds.

    > **Note:** Referencing `r`, `g`, and `b` values inside a `color()` function with a XYZ-based colorspace, `x`, `y`, and `z` values inside a `color()` function with an RGB-based colorspace, or any other characters, is invalid. The origin color channel values available inside the function must match the specified type of colorspace.

> - The `alpha` channel is resolved to a `<number>` between `0` and `1`, inclusive.

Reviewer: oh, you did include it. i'll leave my original comment for the placement and the wording. I also added "alpha" to the introductory sentence, which is might be an additional option.

When defining a relative color, the different channels of the output color can be expressed in several different ways. Below, we'll study some examples to illustrate these.

> It is important to bear in mind that, in the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative `color()` syntax.

Reviewer: ```suggestion
In the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative `color()` syntax.
```
The "important to bear in mind" makes it sound like it's an important point to remember. This is more of an aside, so suggest removing the intro phrasing.

Author: Updated, on all pages.

Let's start with an origin color of `hsl(0 100% 50%)` (equivalent to `red`). While you are unlikely to ever write the following functions because they output the same color as the origin color, this demonstrates how to use the origin color channel values as the output channel values:

```css
color(from hsl(0 100% 50%) srgb r g b)
color(from hsl(0 100% 50%) xyz x y z)
```

> > **Note:** As mentioned above, if the output color is using a different color model to the origin color, the origin color is converted to the same model as the output color in the background so that it can be represented in a way that is compatible (i.e. using the same channels). For example, the {{cssxref("color_value/hsl", "hsl()")}} color `hsl(0 100% 50%)` is converted to `color(srgb 1 0 0)` in the first case above and `color(xyz 0.412426 0.212648 0.0193173)` in the second case.

Reviewer: This is SUPER helpful, and really makes the example on line 114 understandable.
I think it may be even more helpful to move this note down closer to the example where it is first really helpful, maybe to line 120.

Author: OK, I've moved this note down to where you've suggested and made the equivalent change on all other color function pages.

slug: Web/CSS/color_value/color
page-type: css-function
browser-compat: css.types.color.color
spec-urls:

>   - https://drafts.csswg.org/css-color-5/#relative-color-function

Reviewer: Regarding line 4: Should we add `color()` to the sidebar under functions? Yari only does top-level css-function pages under function. How can we get it to go deeper?

>   - https://drafts.csswg.org/css-color-5/#relative-color-function

Author: I've added this to the spec urls list, but I've also kept the link to the #relative-color-function section of the spec. I did this because:

1. I think it is useful to link to precisely where relative colors are specified
2. This keeps it consistent with the other color function pages, which have a spec link to their relative color specifications.

I have no idea how to update the CSSRef macro to add sidebar levels. I remember talk of how horribly complicated that macro is from years gone by, so I don't really want to touch it.

>   - https://drafts.csswg.org/css-color-5/#relative-color-function

Reviewer: This is resolved in https://github.com/mdn/yari/pull/10679. Yay!

## Syntax

```css
/* Absolute values */

color(display-p3 1 0.5 0);
color(display-p3 1 0.5 0 / .5);

/* Relative values */

/* Add a semi-transparent alpha channel to green */
color(from green xyz-d50 x y z / 0.5)
/* Create a magenta color from blue by setting the output
   color's r and b channel values equal to modified
   versions of the origin color's r and b values */
color(from blue display-p3 calc(r + 0.75) g calc(b - 0.35))
```

### Values

Below we have provided descriptions of the allowed values for both absolute and relative colors.

> **Note:** See [Missing color components](/en-US/docs/Web/CSS/color_value#missing_color_components) for the effect of `none`.

#### Absolute values

Functional notation of absolute values: `color(colorspace c1 c2 c3[ / A])`

- `colorspace`

  - : An {{CSSXref("&lt;ident&gt;")}} denoting one of the predefined colorspaces: `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, `rec2020`, `xyz`, `xyz-d50`, or `xyz-d65`.

- `c1`, `c2`, `c3`

>   - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none`. These values represent the component values for the colorspace. When using a `<number>` value, 0-1 represents the bounds of the colorspace. Values outside of that range are permitted, but will be out of {{glossary("Gamut", "gamut")}} for the given colorspace. When using a percentage value, 100% represents 1 and 0% represents 0.

Reviewer 2: `none` should be documented and explained in the context of interpolation because that is where it is easily observable what the difference is between `0` and `none`.

In short, when interpolating between two channels with values `0` and `100` the outcome is `50`, but interplating between `none` and `100` is `100`. When a color value will be used for display and still contains `none` it will behave as `0`.

For relative color, which is already a complex subject, I think it is ok not to explain `none` and only link out to other docs.

>   - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none`. These values represent the component values for the colorspace. When using a `<number>` value, 0-1 represents the bounds of the colorspace. Values outside of that range are permitted, but will be out of {{glossary("Gamut", "gamut")}} for the given colorspace. When using a percentage value, 100% represents 1 and 0% represents 0.

Author: See above

- `A` {{optional_inline}}

  - : An {{CSSXref("&lt;alpha-value&gt;")}} or the keyword `none`, where the number `1` corresponds to `100%` (full opacity).

#### Relative values

Functional notation of relative values: `color(from color colorspace c1 c2 c3[ / alpha])`

- `from`
  - : The keyword `from` is always included when defining a relative color.
- `color`

>   - : A {{cssxref("&lt;color&gt;")}} value representing the **origin color** that the relative color is based on. Note that this can be _any_ valid {{cssxref("&lt;color&gt;")}} syntax, including another relative color.

Author: Yup, this is a good call — I've done this on all color function pages.

- `colorspace`
  - : An {{CSSXref("&lt;ident&gt;")}} denoting one of the predefined colorspaces: `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, `rec2020`, `xyz`, `xyz-d50`, or `xyz-d65`.
- `c1`, `c2`, `c3`

>   - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none`. These values represent the component values for the output color. When using a `<number>` value, 0-1 represents the bounds of the colorspace. Values outside of that range are permitted, but will be out of {{glossary("Gamut", "gamut")}} for the given colorspace. When using a percentage value, 100% represents 1 and 0% represents 0.

Author: OK, I've updated all color function pages to make it clear what the value of `none` is in every case.

Bear in mind that saying they can be written as `r g b` or `x y z` is a bit misleading — those values are made available inside the function for the developer to use to define the component values if they choose to, but the values themselves are still going to be numbers, percentages, or none. 

This is why I include a separate section below to explain the origin color channel values and how they can be used when specifying the output color channel values. This is one of the most confusing parts of this syntax.

Here, I don't want to start getting into whether a color specified with relative color syntax is a relative color, depending on whether you specify different output channel values or not. It is still relative color syntax, regardless of that, and I think it will just confuse people more. Again, I cover that notion in the following section.

- `alpha` {{optional_inline}}
  - : An {{CSSXref("&lt;alpha-value&gt;")}} or the keyword `none`, where the number `1` corresponds to `100%` (full opacity). This represents the alpha channel value of the output color. If the `alpha` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color.

#### Defining relative color output channel components

> The browser destructures the origin color into `color()` component values for the specified colorspace that are made available inside the function as three color channel values, and `alpha`. These can be used in defining the output color channel values if desired, and take the following values:

Author: OK, how about this:

> When using relative color syntax inside a `color()` function, the browser first converts the origin color into an equivalent representation in the specified colorspace. It then separates that color into three different color channel values plus an alpha channel value. These channel values are made available inside the function to be used when defining the output color channel values:

> - The three color channel values are resolved to a `<number>`, nominally between 0 and 1, but as explained above they may be outside these bounds. Depending on the specified colorspace, these values will be one of the following:

Author: Grammarly doesn't seem to hate the original, and I'd rather not put more commas into this sentence — too many commas.

>   - `r`, `g`, and `b`: Color channel values for RGB-based colorspaces, that is `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, and `rec2020`.

Author: Good call, updated.

>   - `x`, `y`, and `z`: Color channel values for CIE XYZ-based colorspaces, that is `xyz`, `xyz-d50`, and `xyz-d65`.

Author: Updated.

>     > **Note:** As per spec, specifying `r`, `g`, and `b` values inside a `color()` function with a CIE XYZ-based colorspace, or specifying `x`, `y`, and `z` values inside a `color()` function with an RGB-based colorspace, will result in an invalid color value. The origin color values have to match the specified type of colorspace.

Author: I've used some of your wording here, but I rewrote this bit anyway because of the confusion it caused earlier. Have another look and see what you think.

> - The `alpha` channel is resolved to a `<number>` between 0 and 1 which represents the origin color's alpha value.

Author: Yup, I thought it made sense.

> When defining a relative color, the different channels of the output color can be expressed in several different ways.

Author: Yeah, but `from` is part of the actual syntax; "to" isn't. I'm a bit concerned that this might make people think they've missed some syntax when they haven't. I did make it clear what we are talking about expressing in the second half of the sentence.

For example, let's start with an origin color of `hsl(0 100% 50%)` (equivalent to `red`). The following functions output the same color as the origin color — they use the origin color channel values as the output channel values:

```css
color(from hsl(0 100% 50%) srgb r g b)
color(from hsl(0 100% 50%) xyz x y z)
```

> > **Note:** If the output color is using a different color model to the origin color, the origin color is converted to the same model as the output color in the background so that it can be represented in a way that is compatible (i.e. using the same channels).

Author: Ah, so if you like the wording suggestion I provided in the earlier comment, we can probably get rid of this note. Let me know what you think.

> These functions use absolute values for the output color channel values, outputting completely different colors not based on the origin color:

Author: See below comment

```css
color(from hsl(0 100% 50%) srgb 0.749938 0 0.609579)
color(from hsl(0 100% 50%) xyz 0.75 0.6554 0.1)
```

> > **Note:** Bear in mind that, if you are using relative color syntax but outputting the same color as the origin color or a color not based on the origin color at all, you are not really creating a relative color. You'd be unlikely to ever do this in a real codebase, and would probably just use an absolute color value instead. But, we felt it useful to explain that you _can_ do this with relative color syntax, as a starting point for learning about it.

Author: I felt like including the above explanations you added, AND making this note even more verbose, seemed to make things a bit more awkward and confusing, even though I appreciate why you want to make the point this is all making earlier on (and I do think that's a good idea).

I've instead chosen to rewrite the note and include it as normal text before the first two examples, like so:

It is important to bear in mind that, in the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative color syntax.

WDYT?

> > **Note:** Bear in mind that, if you are using relative color syntax but outputting the same color as the origin color or a color not based on the origin color at all, you are not really creating a relative color. You'd be unlikely to ever do this in a real codebase, and would probably just use an absolute color value instead. But, we felt it useful to explain that you _can_ do this with relative color syntax, as a starting point for learning about it.

Author: @romainmenke I think you are correct technically, and this is why I didn't want to start mentioning this stuff in the syntax/values section.

However, I think @estelle's point about people reading this and thinking "why would i ever do this" before then seeing the note is a good one. At worst, it might make them think that the relative colors syntax is pointless or bad, because it seems like a more complex way of doing things without much benefit, or because the colors are not really relative. This way, it is clearer that these examples are only for illustration.

> > **Note:** Bear in mind that, if you are using relative color syntax but outputting the same color as the origin color or a color not based on the origin color at all, you are not really creating a relative color. You'd be unlikely to ever do this in a real codebase, and would probably just use an absolute color value instead. But, we felt it useful to explain that you _can_ do this with relative color syntax, as a starting point for learning about it.

Reviewer 2: I meant omitting everything about this aspect, the note and the examples :)

> > **Note:** Bear in mind that, if you are using relative color syntax but outputting the same color as the origin color or a color not based on the origin color at all, you are not really creating a relative color. You'd be unlikely to ever do this in a real codebase, and would probably just use an absolute color value instead. But, we felt it useful to explain that you _can_ do this with relative color syntax, as a starting point for learning about it.

Author: I think we do need to include some examples and explanation of the different ways in which users can set the output color channel values. And also, what the origin color channel values made available by the browser are. I found this bit confusing when first learning this syntax.

> > **Note:** Bear in mind that, if you are using relative color syntax but outputting the same color as the origin color or a color not based on the origin color at all, you are not really creating a relative color. You'd be unlikely to ever do this in a real codebase, and would probably just use an absolute color value instead. But, we felt it useful to explain that you _can_ do this with relative color syntax, as a starting point for learning about it.

Reviewer 2: I don't think that these two are special cases:
- all channels have numbers unrelated to the origin color
- all channels are taken exactly from the origin color and in the same order

I think calling these out specifically makes it more confusing.
But I might be wrong about that :)

For me at least these two cases are the same as any other way of using relative color syntax. Numbers go in, a color comes out.

The following functions use two of the origin color channel values for the output color channel values, but use a new value for the other output channel value, creating a relative color based on the origin color in each case:

```css
color(from hsl(0 100% 50%) srgb r 1 b)
color(from hsl(0 100% 50%) xyz x y 0.5)
```

The following functions use the origin color channel values inside {{cssxref("calc")}} functions to calculate new channel values for the output colors:

```css

> color(from hsl(0 100% 50%) srgb calc(r - 0.4) calc(g + 0.1) calc(b + 0.6))

Author: When you say "provide the output values", what do you mean? Show what the equivalent absolute color would be? This is fairly tricky in some cases, as the computed style is always the equivalent `color()` value regardless of the function being used (at least, it is in Chrome).

Or do you mean, show a visual output of the color?

> color(from hsl(0 100% 50%) srgb calc(r - 0.4) calc(g + 0.1) calc(b + 0.6))

Reviewer: in this case, based on:

> color(from hsl(0 100% 50%) srgb calc(r - 0.4) calc(g + 0.1) calc(b + 0.6))

my math indicates:

hsl(0 100% 50%) = rgb(255 0 0) = rgb(100% 0% 0%)
r= 255 or 1
calc(r - 0.4) = 0.6 = 60%
calc(g + 0.1) = 0.1 = 10%
calc(b+0.6) = 0.6 = 60%

so the "output" is rgb(60% 10% 60%)

```
color(from hsl(0 100% 50%) srgb calc(r - 0.4) calc(g + 0.1) calc(b + 0.6)) /* the color equals rgb(60% 10% 60%)

> color(from hsl(0 100% 50%) srgb calc(r - 0.4) calc(g + 0.1) calc(b + 0.6))

Author: You've got the calculated percentages correct, but the computed output colors are `color()` functions. I've added comments to the code as follows:

```css
color(from hsl(0 100% 50%) srgb calc(r - 0.4) calc(g + 0.1) calc(b + 0.6))
/* Computed output color: color(srgb 0.6 0.1 0.6)  */

color(from hsl(0 100% 50%) xyz calc(x - 0.3) calc(y + 0.3) calc(z + 0.3))
/* Computed output color: color(xyz-d65 0.112426 0.512648 0.319317) */
```

Do you think that works?

> color(from hsl(0 100% 50%) srgb calc(r - 0.4) calc(g + 0.1) calc(b + 0.6))

Author: For the moment, I've just added this to the `color()` page, but if you think this is good, I'll do the same update on the other places too.

## Syntax

```css
/* Absolute values */

> 

Author: I think you're right, and I also think this is more consistent with how we usually write comments relative to the code they are commenting on. I've made this change and made sure it is done consistently on all the other color function pages too.

color(display-p3 1 0.5 0);
color(display-p3 1 0.5 0 / .5);

/* Relative values */

color(from green srgb r g b / 0.5)
color(from #0000FF xyz calc(x + 0.75) y calc(z - 0.35))
```

### Values

Below are descriptions of the allowed values for both absolute and [relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors).

#### Absolute value syntax

```text
color(colorspace c1 c2 c3[ / A])
```

The parameters are as follows:

- `colorspace`

  - : An {{CSSXref("&lt;ident&gt;")}} denoting one of the predefined colorspaces: `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, `rec2020`, `xyz`, `xyz-d50`, or `xyz-d65`.

- `c1`, `c2`, `c3`

>   - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none` (equivalent to `0` in this case). These values represent the component values for the colorspace. When using a `<number>` value, 0-1 represents the bounds of the colorspace. Values outside of that range are permitted but will be out of {{glossary("Gamut", "gamut")}} for the given colorspace. When using a percentage value, 100% represents 1 and 0% represents 0.

Author: All looks good to me. I've made this update, and also made the equivalent update in the relative colors syntax section.

>   - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none` (equivalent to `0` in this case). These values represent the component values for the colorspace. When using a `<number>` value, 0-1 represents the bounds of the colorspace. Values outside of that range are permitted but will be out of {{glossary("Gamut", "gamut")}} for the given colorspace. When using a percentage value, 100% represents 1 and 0% represents 0.

Reviewer: @chrisdavidmills This is what you asked me to highlight in the Slack DM

- `A` {{optional_inline}}

>   - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `1` corresponds to `100%` (full opacity), or the keyword `none` to explicitly specify no alpha channel. If included, the value is preceded by a slash (`/`).

Reviewer: ```suggestion
  - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` to explicitly specify no alpha channel. If included, the value is preceded by a slash (`/`).
```

> **Note:** See [Missing color components](/en-US/docs/Web/CSS/color_value#missing_color_components) for more information on the effect of `none`.

#### Relative value syntax

```text
color(from <color> colorspace c1 c2 c3[ / A])
```

The parameters are as follows:

> - `from <color>`
>   - : The keyword `from` is always included when defining a relative color, followed by a {{cssxref("&lt;color&gt;")}} value representing the **origin color**. This is the original color that the relative color is based on. Note that the origin color can be _any_ valid {{cssxref("&lt;color&gt;")}} syntax, including another relative color.
> - `colorspace`
>   - : An {{CSSXref("&lt;ident&gt;")}} denoting one of the predefined colorspaces: `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, `rec2020`, `xyz`, `xyz-d50`, or `xyz-d65`.
> - `c1`, `c2`, `c3`
>   - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none` (equivalent to `0` in this case). These values represent the component values for the output color. When using a `<number>` value, 0-1 represents the bounds of the colorspace. Values outside of that range are permitted but will be out of {{glossary("Gamut", "gamut")}} for the given colorspace. When using a percentage value, 100% represents 1 and 0% represents 0.
> - `A` {{optional_inline}}
>   - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `1` corresponds to `100%` (full opacity), or the keyword `none` to explicitly specify no alpha channel. This represents the alpha channel value of the output color. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).

Reviewer: ```suggestion
- `from <color>`
    - : The keyword `from` is always included when defining a relative color, followed by a {{cssxref("&lt;color&gt;")}} value representing the **origin color**. This is the original color the relative color is based on. Note that the origin color can be _any_ valid {{cssxref("&lt;color&gt;")}} syntax, including another relative color.
- `colorspace`
  - : An {{CSSXref("&lt;ident&gt;")}} denoting the {{glossary("color space")}} of the output color, generally one of the predefined color spaces: `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, `rec2020`, `xyz`, `xyz-d50`, or `xyz-d65`.
- `c1`, `c2`, `c3`
  - : Each value can be written as a {{CSSXref("number")}}, a {{CSSXref("percentage")}}, or the keyword `none` (equivalent to `0` in this case). These values represent the component values for the output color. When using a `<number>` value, generally `0` to `1` represents the bounds of the colorspace. Values outside of that range are permitted but will be out of {{glossary("Gamut", "gamut")}} for the given colorspace. Generally, when using a percentage value, `100%` represents `1` and `0%` represents `0`.
- `A` {{optional_inline}}
  - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` to `100%` (full opacity), or the keyword `none` to explicitly specify no alpha channel. This represents the alpha channel value of the output color. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).
```

Author: I've mostly used these changes. I wasn't convinced by "generally one of the predefined color spaces", as it impoies you can define your own custom color spaces, which I'm pretty sure is not possible.

#### Defining relative color output channel components

When using relative color syntax inside a `color()` function, the browser converts the origin color into an equivalent color in the specified colorspace (if it is not already specified as such). The color is defined as three distinct color channel values plus an alpha channel value (`alpha`). These channel values are made available inside the function to be used when defining the output color channel values:

> - The three color channel values of the origin color are resolved to a `<number>`. Depending on the specified colorspace, these values will be one of the following:

Author: I have an issue with this suggestion. Earlier on in the syntax sections, we say the values are predefined. Repeating it here therefore seems unnecessary, IMO, plus also I think that at worst, it could imply that there is some kind of way to define and specify your own custom (non-predefined) color spaces, which is not possible.

> - The three color channel values of the origin color are resolved to a `<number>`. Depending on the specified colorspace, these values will be one of the following:

Author: So, I'm leaving this one for now.

> - The three color channel values of the origin color are resolved to a `<number>`. Depending on the specified colorspace, these values will be one of the following:

Reviewer: @chrisdavidmills This is what you asked me to highlight in the Slack DM

  - `r`, `g`, and `b`: Color channel values for the RGB-based colorspaces `srgb`, `srgb-linear`, `display-p3`, `a98-rgb`, `prophoto-rgb`, and `rec2020`.
  - `x`, `y`, and `z`: Color channel values for the CIE XYZ-based colorspaces `xyz`, `xyz-d50`, and `xyz-d65`.

>     > **Note:** Each of these values is usually between 0 and 1 but, as explained above, they may be outside these bounds.

Author: Updated. I've also done a scan through all the other articles to catch other non-code instances of number values.

>     > **Note:** Referencing `r`, `g`, and `b` values inside a `color()` function with a XYZ-based colorspace, or referencing `x`, `y`, and `z` values inside a `color()` function with an RGB-based colorspace, is invalid. The origin color channel values available inside the function will match the specified type of colorspace.

Author: Updated

> - The `alpha` channel is resolved to a `<number>` between 0 and 1.

Reviewer: ```suggestion
- The `alpha` channel is resolved to a `<number>` between `0` and `1`, inclusive.
```

Author: Updated

When defining a relative color, the different channels of the output color can be expressed in several different ways. Below, we'll study some examples to illustrate these.

> It is important to bear in mind that, in the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative color syntax.

Reviewer: ```suggestion
It is important to bear in mind that, in the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative `color()` syntax.
```

Author: updated

Let's start with an origin color of `hsl(0 100% 50%)` (equivalent to `red`). While you are unlikely to ever write the following functions because they output the same color as the origin color, this demonstrates how to use the origin color channel values as the output channel values:

```css
color(from hsl(0 100% 50%) srgb r g b)
color(from hsl(0 100% 50%) xyz x y z)

> ```

Reviewer: we should likely include an example where we transition the A value.

Author: I agree with you here. I thought about this quite a bit — if I included it here, then it would make sense to include an alpha channel version of all the other examples, which would be consistent but become very repetitious and long-winded.

Instead, I elected to include a small section further down that discusses relative colors that include alpha channels, explains what the deal is with default values as a reminder (as we only mention it briefly earlier on), and provides a couple of examples.

I've added the same equivalent section to all the color function pages. Hope you like it ;-)

```html hidden
<div id="container">
  <div class="item" id="one"></div>
  <div class="item" id="two"></div>
  <div class="item" id="three"></div>
</div>
```

#### CSS

```css hidden
#container {
  display: flex;
  width: 100vw;
  height: 100vh;
  box-sizing: border-box;
}

.item {
  flex: 1;
  margin: 20px;
}
```

```css
:root {
  --base-color: orange;
}

#one {
  background-color: color(
    from var(--base-color) display-p3 r calc(g + 0.15) calc(b + 0.15)
  );
}

#two {
  background-color: var(--base-color);
}

#three {
  background-color: color(
    from var(--base-color) display-p3 r calc(g - 0.15) calc(b - 0.15)
  );
}

/* Use @supports to add in support for Safari 16.4+, which supports old
   syntax that requires r g b values to be specified as percentages (with
   units) in calculations */
@supports (color: color(from red display-p3 r g calc(b + 30%))) {

>   #one {

Reviewer: Invert text so you explain what is being done for "old syntax", then mention that it is supported in safari.
In other words, explain generically, with the mention of a specific browser having a quirk as the least important and therefor last part (or omitted)

Also, do browsers not support having a % in a calc() within a relative color? if they do, this will overwrite the values for all the browsers. Generally, we put hacks first, with current best practices taking precedence (coming later in source code order) over fallbacks. alternatively, do `calc(g * 115%)` in line 340, or would that not work in Safari

Author: I've inverted the text as I think that makes sense. I'm not sure about putting the incorrect screwed-up code first. It makes sense in production code, but this is a demo, and it is good to show readers the standard code first, IMO. It'll also be good for future-proofing when Safari fixes it, up until we remember to remove this bit ;-)

I tried `calc(g * 115%)` for the sake of argument. That doesn't seem to work in Saf or Chrome.

Reviewer 2: Pretty sure that this is a bug.
https://github.com/w3c/csswg-drafts/issues/10041

{{EmbedLiveSample("using_color-gamut_media_queries_with_color")}}

### Using relative colors with color()

> The below example styles three {{htmlelement("div")}} elements with different background colors. The middle one is given the unmodified `--base-color`, while the left and right ones are given lightened and darkened variants of that `--base-color`.

Reviewer: ```suggestion
This example styles three {{htmlelement("div")}} elements with different background colors. The middle one is given the unmodified `--base-color`, while the left and right ones are given lightened and darkened variants of that `--base-color`.
```

Author: Updated, on all color function pages.

{{EmbedLiveSample("using_color-gamut_media_queries_with_color")}}

### Using relative colors with color()

This example styles three {{htmlelement("div")}} elements with different background colors. The middle one is given the unmodified `--base-color`, while the left and right ones are given lightened and darkened variants of that `--base-color`.

> These variants are defined using relative colors — the `--base-color` [custom property](/en-US/docs/Web/CSS/--*) is passed into a `color()` function, and the output colors have their `g` and `b` channels modified to achieve the desired effect via `calc()` functions. The lightened color has 15% added to those channels, and the darkened color has 15% subtracted from those channels.

Author: I am totally cool with the description edit — that is super useful.

As for this comment, ummm, when I look at it from left to right I immediately think lighter and darker. I don't know how else to express it, and I don't know if it's a great use of my time. I'll get a second opinion on it and return to this one.

## See also

- [The `<color>` data type](/en-US/docs/Web/CSS/color_value) for a list of all color notations
- [Using relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors)

> - [CSS colors](/en-US/docs/Web/CSS/CSS_colors/) module

Reviewer 3: ```suggestion
- [CSS colors](/en-US/docs/Web/CSS/CSS_colors) module
```
In other 7 docs as well. It'll resolve 8 flaws.

Author: Thanks Onkar! I've updated this in all 8 docs.
